Preload with indexedDB and memory cache

kyon0304 opened the pull request.

kyon0304 committed:
add indexedDB wrapper store.js and use it in loader.js

kyon0304 committed:
using memory cache

kyon0304 committed:
use indexedDB and memory cache to manange download image

kyon0304 committed:
Merge branch 'master' into preload

Conflicts:
	src/manager/loader.js

kyon0304 committed:
Merge branch 'bugfix' into preload

kyon0304 committed:
Merge branch 'master' into preload

Conflicts:
	src/manager/loader.js

kyon0304 commented:
cc @hxgdzyuyi

hxgdzyuyi commented on src/mods/store.js:
```diff
@@ -0,0 +1,109 @@
+//import app from 'app'
```
`mods` 文件夹通常用来存放与 具体 UI 逻辑无关的，高重用的组件。 比如 mods/cookie mods/browser @kyon0304

kyon0304 replied:
@hxgdzyuyi 嗯.. 窝是看 request 放在这里于是也放这儿了, 那窝把它移到 `manager/`下面?

hxgdzyuyi commented on src/manager/loader.js:
```diff
@@ -88,22 +88,45 @@ class Loader {
-    this.store(page, val)
+    let key = this.model.getCurrentPage()
+
+    if(this.presentPages.has(key)) { return }
```
没有判断 indexedDB 中是否存在？

kyon0304 replied:
@hxgdzyuyi 在 storeImage 中做的处理, 如果已经存在, 那么 add 操作就会 reject

kyon0304 committed:
change store api according to localforage

kyon0304 committed:
move store.js from mods/ to manager/

kyon0304 commented:
cc @hxgdzyuyi

hxgdzyuyi commented on src/manager/loader.js:
```diff
@@ -33,10 +39,16 @@ function spawn(fn) {
     this.THRESHOLD = 5
     this.xhr = undefined
+    this.presentPages = new Map()
+    let config = {
```
加个 //TODO(kyon): balabala 之类的

kyon0304 replied:
@hxgdzyuyi 不懂, todo 啥?

hxgdzyuyi replied:
@kyon0304 version 1 是啥。。。

kyon0304 replied:
@hxgdzyuyi indexeddb 只有在 versionchange 的时候才会更新数据库表, 第一次创建也是, 所以需要 version

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,184 @@
+        db: null
+      }
+
+    for (let i in options) {
```
用 Object.assign ?

kyon0304 replied:
@hxgdzyuyi 嗯好

kyon0304 replied:
@hxgdzyuyi fixed in commit b321eca

kyon0304 committed:
use Object.assign instead of for..in

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,184 @@
+            reject('not found')
+          }
+        }
+        req.onerror = (evt) => {
```
统一 `evt` 用 `e` 来表示吧。

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,184 @@
+    })
+  }
+
+  config(options) {
```
这个函数有用到？

hxgdzyuyi commented on src/manager/loader.js:
```diff
@@ -79,31 +91,50 @@ class Loader {
       , noop = function() {}
+      , self = this
+
+    _.once(this.store.config({ 'storeName': encodeURI(this.model.getBookTitle()) }))
```
这个不是在开头设定过一次了么。 为什么还需要再设定一次啊。

kyon0304 replied:
@hxgdzyuyi 那个是默认的, 这个是第一次打开数据库的之前, 先把表名改成书名.

kyon0304 replied:
@hxgdzyuyi 在 constructor 运行的时候, book model 还没有被创建, 所以只能把这个配置往后放.

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+  }
+
+  config(options) {
+    if (typeof(options) === 'object') {
```
这个可以用 _.isObject() 来判断

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+        db: null
+      }
+
+    Object.assign(dbInfo, options)
```
这里为什么不用 this.config() 函数呢？

kyon0304 replied:
@hxgdzyuyi this.config() 是配置 this._config 参数的, 这个是配置 this._dbInfo 的

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+    return ready
+  }
+
+  getObjectStore(store_name, mode) {
```
这个 store_name 为何不能直接用 this.config.storeName

hxgdzyuyi replied:
这个 store_name 为何不能直接用 this.config.storeName

kyon0304 replied:
@hxgdzyuyi 这个参数可以是一个数组, 虽然现在还没有用到.

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+    let self = this
+
+    let ready = new Promise((resolve, reject) => {
+      if (self._ready === null) {
```
_initStorage 是否可以和 ready 函数合并成 ready 函数？

kyon0304 replied:
@hxgdzyuyi 应该可以的, 不过两个函数看上去是不是清晰一些?

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+          if (evt.target.result) {
+            resolve(evt.target.result.imageBlob)
+          } else {
+            reject('not found')
```
reject 不统一啊， 一个是 errorCode 一个是 message。

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+            let value = cursor.value
+              , result = iterator(value, cursor.key, iterationNumber++)
+
+            if ( result !== void(0)) {
```
_.isUndefined

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,182 @@
+    })
+  }
+
+  iterate(iterator) {
```
没用到？

kyon0304 replied:
@hxgdzyuyi fixed in commit bf5f15d

hxgdzyuyi commented:
@kyon0304  好了

kyon0304 committed:
fix according to PR comments, add backbone events when book model is ready to create indexeddb and initiate memory map

kyon0304 commented:
cc @hxgdzyuyi

kyon0304 commented on src/manager/loader.js:
```diff
@@ -19,24 +21,42 @@ var Model = Backbone.Model.extend({
     this.xhr = undefined
+    this.presentPages = new Map()
+    //XXX(kyon) triggered twice
+    app.on('create: book', () => {
```
@hxgdzyuyi 这个地方不懂为什么会触发两次, 求教

hxgdzyuyi replied:
@kyon0304 你可以用 `debugger` 来断点，看看调用链

kyon0304 committed:
Bugfix backbone model event trigger twice

kyon0304 commented:
cc @hxgdzyuyi

kyon0304 committed:
use src instead of page

kyon0304 committed:
bugfix DRY

kyon0304 commented:
cc @hxgdzyuyi 改完了, 帮忙 review 吧~

hxgdzyuyi commented on src/main.js:
```diff
@@ -19,6 +19,7 @@ $('body').prepend(appViewWrapper)
 
 bookModel.fetch({ url: '/content.json' })
   .done(() => {
+    app.trigger('create:book')
```
这个应该移到 app.createModel 里头去吧 @kyon0304

kyon0304 replied:
@hxgdzyuyi  但是 app.createModel 的时候, content.json 还没有下载下来阿

hxgdzyuyi replied:
@kyon0304  换个名字吧 fetched:book 之类的吧

hxgdzyuyi commented on src/manager/loader.js:
```diff
@@ -19,24 +21,46 @@ var Model = Backbone.Model.extend({
+  }
+
+  initPresentPages() {
+    return this.store.iterate((val, key, iterationNumber) => {
```
为什么一定要在内存里头留一份啊。 未来会去掉么 cc @kyon0304

kyon0304 replied:
@hxgdzyuyi 不会去掉吧, 因为数据库操作都是异步的, 在内存里面留一份方便判断是否已加载, 以及方便获取某张图片.

hxgdzyuyi commented on src/manager/store.js:
```diff
@@ -0,0 +1,180 @@
+    })
+  }
+
+  config(options) {
```
这个函数没用过？ @kyon0304

kyon0304 replied:
@hxgdzyuyi 嗯, 后来判断出 book model 创建时机以后, 这个函数就没有用到了. 要删掉吗?

hxgdzyuyi replied:
肯定啊

kyon0304 replied:
@hxgdzyuyi fixed in commit a8d587c

kyon0304 committed:
remove dead code

kyon0304 committed:
Bugfix

1. 将 preload 中的 break 判断提到 continue 前, 防止页数超出总页数限制
2. store.js 中 iterate 函数修改 resolve 判断条件, 以及 typo 修改
3. 内存 map 的 key 遗留为 page 的改为 url
4. 箭头函数中使用 self 暂存外部的 this
5. 修改 Model 的 getCurrentImageUri() 函数符合 DRY

kyon0304 committed:
fix according to PR comments

1. rename presentPages to cachedPages
2. rename create:book to fetched:book

kyon0304 committed:
bugfix page to url

kyon0304 committed:
不同的书对应数据库不同版本

不同的书通过 uuid 辨别
Ps. bugfix book model 定义方式改变带来的问题

Conflicts:
	src/manager/loader.js
	src/models/book.js

kyon0304 committed:
Merge branch 'master' of github.com:komic-awesome/komic into preload

Conflicts:
	src/models/book.js

kyon0304 committed:
[bugfix] return promise object

kyon0304 committed:
add uuid in store name
